pipeline_dp/accumulator.py

```diff
@@ -0,0 +1,117 @@
+import abc
+import typing
+import pickle
+from functools import reduce
+
+
+def merge(accumulators: typing.Iterable[
+  'Accumulator'] = None) -> 'Accumulator':
+  """Merges the accumulators.
+
+  Args:
+    accumulators:
+
+  Returns: Accumulator instance with merged values.
+  """
+  unique_accumulator_types = {type(accumulator).__name__ for accumulator in
+                              accumulators}
+  if len(unique_accumulator_types) > 1:
+    raise TypeError(
+      "Accumulators should all be of the same type. Found accumulators of "
+      + f"different types: ({','.join(unique_accumulator_types)}).")
+
+  return reduce(lambda acc1, acc2: acc1.add_accumulator(acc2), accumulators)
+
+
+class Accumulator(abc.ABC):
+  """Base class for all accumulators.
+
+    Accumulators are objects that encapsulate aggregations and computations of
+    differential private metrics.
+  """
+
+  @abc.abstractmethod
+  def add_value(self, value):
+    """Adds the value to each of the accumulator.
+    Args:
+      value: value to be added.
+
+    Returns: self.
+    """
+    pass
+
+  @abc.abstractmethod
+  def add_accumulator(self, accumulator: 'Accumulator') -> 'Accumulator':
+    """Merges the accumulator to self and returns self.
+      Args:
+        accumulator:
+
+      Returns: self
+    """
+    pass
+
+  @abc.abstractmethod
+  def compute_metrics(self):
+    pass
+
+  def serialize(self):
+    return pickle.dumps(self)
+
+  @classmethod
+  def deserialize(cls, serialized_obj: str):
+    deserialized_obj = pickle.loads(serialized_obj)
+    if not isinstance(deserialized_obj, cls):
+      raise TypeError("The deserialized object is not of the right type.")
+    return deserialized_obj
+
+
+class CompoundAccumulator(Accumulator):
+  """Accumulator for computing multiple metrics.
+
+    CompoundAccumulator contains one or more accumulators of other types for
+    computing multiple metrics.
+    For example it can contain [CountAccumulator,  SumAccumulator].
+    CompoundAccumulator delegates all operation to the internal accumulators.
+  """
+
+  def __init__(self, accumulators: typing.Iterable['Accumulator']):
+    self.accumulators = accumulators
+
+  def add_value(self, value):
+    for accumulator in self.accumulators:
+      accumulator.add_value(value)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+    'CompoundAccumulator':
+    """Merges the accumulators of the CompoundAccumulators.
+    The expectation is that the input accumulators are of the same type and
+    are in the same order."""
+
+    if len(accumulator.accumulators) != len(self.accumulators):
+      raise ValueError(
+        "Accumulators in the input are not of the same size."
+        + f" Expected size = {len(self.accumulators)}"
+        + f" received size = {len(accumulator.accumulators)}.")
+
+    expected_type_order = ",".join([type(accumulator).__name__ for
```

Comment:
Nit: Checking of types might be done in slighly shorter and more efficient way (no need to create lists):

```
for acc1, acc2 in zip(..., ...):
  if type(acc1) is not type(acc2):
     raise ....        # here we can just provide the type of the 1st mismatched accumulators 
```

Reply:
Got it! Let me know if this looks good!